core/src/main/java/org/apache/calcite/rex/RexCall.java

```diff
@@ -119,36 +122,178 @@ public RelDataType getType() {
     // "c IS NOT NULL" occurs when we expand EXISTS.
     // This reduction allows us to convert it to a semi-join.
     switch (getKind()) {
+    case IS_NULL:
+    case IS_UNKNOWN:
+      return operands.get(0).isAlwaysNull();
     case IS_NOT_NULL:
+      // IS_NOT_UNKNOWN does not exist yet
       return !operands.get(0).getType().isNullable();
+    case IS_TRUE:
     case IS_NOT_FALSE:
+    case MAX:
+    case MIN:
+      return operands.get(0).isAlwaysTrue();
     case NOT:
       return operands.get(0).isAlwaysFalse();
     case IS_NOT_TRUE:
     case IS_FALSE:
+      return operands.get(0).isAlwaysFalse();
     case CAST:
       return operands.get(0).isAlwaysTrue();
+    case AND:
+      for (RexNode operand : operands) {
+        if (!operand.isAlwaysTrue()) {
+          return false;
+        }
+      }
+      return true; // all operands are always true => AND is always true
+    case OR:
+      for (RexNode operand : operands) {
+        // OR(true, null) => true
+        if (operand.isAlwaysTrue()) {
+          return true;
+        }
+      }
+      return false;
+    case EQUALS:
+    case NOT_EQUALS:
+    {
+      RexNode o0 = operands.get(0);
+      RexNode o1 = operands.get(1);
+      if (!getType().isNullable() && o0.equals(o1)) {
+        return getKind() == SqlKind.EQUALS;
+      }
+      int k0 = o0.isAlwaysTrue() ? 1 : (o0.isAlwaysFalse() ? 2 : 0);
+      if (k0 != 0) {
+        int k1 = o1.isAlwaysTrue() ? 1 : (o1.isAlwaysFalse() ? 2 : 0);
+        return k1 != 0 && k0 == k1 ^ (getKind() != SqlKind.EQUALS);
+      }
+      return false;
+    }
     default:
       return false;
     }
   }
 
   @Override public boolean isAlwaysFalse() {
     switch (getKind()) {
+    case IS_NOT_NULL:
+      // IS_NOT_UNKNOWN does not exist yet
+      return operands.get(0).isAlwaysNull();
     case IS_NULL:
+    case IS_UNKNOWN:
       return !operands.get(0).getType().isNullable();
+    case IS_FALSE:
     case IS_NOT_TRUE:
     case NOT:
       return operands.get(0).isAlwaysTrue();
     case IS_NOT_FALSE:
     case IS_TRUE:
     case CAST:
+    case MAX:
+    case MIN:
       return operands.get(0).isAlwaysFalse();
+    case OR:
+      for (RexNode operand : operands) {
+        if (!operand.isAlwaysFalse()) {
```

Comment:
I think this is a little bit too much...it will work; but a single isAlwaysTrue() call will do a full visit on the subtree...
As a public method I would think about it as something which takes O(1) time...

I think these cases should be handled by RexSimplify

Reply:
Would `if (operands.size()>3) {return false;}` qualify as `O(1)`?

Reply:
By the way, these "complex" `always*` enable lots of optimizations for `simplify(...)`.
I have just added a static flag to enable-disable these `isAlways...`, and here are some examples:

```
opt1 = AND($4, OR(NOT(false), null, $13, IS NOT DISTINCT FROM(true, false)), $0)
opt2 = AND($4, $0)
```

```
opt1 = AND(OR(IS NULL(null), NOT(null), <=(false, $1), IS NOT DISTINCT FROM(null, null)), OR(null, $4, NOT($4)))
opt2 = OR(null, $4, NOT($4))
```